### DISCUSSION TITLE
How should I route the runoff on each grid to the watershed outlet?

### DISCUSSION THREAD

**Discussion by FredPre** [wschwanghart/topotoolbox#40] [Q&A] [answered]
Dear Wolfgang and Colleagues,
     I have gridded runoff depth data (for example, a 30*30 matrix) based on the hydrological simulation. Then, I want to route the runoff of each grid to the next downstream grid and update the new runoff, until runoff reaches the last grid that serves as the outlet. Thus, I can compare the routed runoff simulations with the observations from the hydrological station located at the outlet.

**Comment by wschwanghart** [comment, marked as answer]
Hi, based on the highly simplified assumption that travel times between adjacent pixels are the same everywhere, we can formulate a very simple runoff model. Here, you have an instantaneous runoff in each grid cell at the beginning of the simulation. If you want to add precipitation (add a source term), you'll need to update the runoff grid accordingly after each time step. Note, however, that the assumption for constant travel times might work on hillslopes, but likely not anymore once the water enters the rivers.

```
DEM = GRIDobj('srtm_bigtujunga30m_utm11.tif');
FD  = FLOWobj(DEM);
RO  = GRIDobj(DEM)+1;
n   = prod(DEM.size);
for r = 1:1000
    % Runoff in each giving cell
    ro = RO.Z(FD.ix);
    % Remove runoff from giving cell
    RO.Z(FD.ix) = RO.Z(FD.ix)-ro;
    % Add runoff to receiving cell
    RO.Z(:,:) = reshape(accumarray(FD.ixc,ro,[n,1]), DEM.size);
    imagesc(RO); 
    pause(.1);
end
```

Alternatively, and in fact much more elegant is to use the transfer matrix M (see Schwanghart and Kuhn, 2010). 

```
M = FLOWobj2M(FD);

RO  = GRIDobj(DEM)+1;
n   = prod(DEM.size);
for r = 1:1000
    RO.Z = reshape(M' * RO.Z(:),DEM.size);
    imagesc(RO); 
    pause(.1);
end
```

References

Schwanghart W, Kuhn NJ. 2010. TopoToolbox: A set of Matlab functions for topographic analysis. Environmental Modelling & Software 25 : 770–781. DOI: [10.1016/j.envsoft.2009.12.002](https://doi.org/10.1016/j.envsoft.2009.12.002)
